lib/java/AdapterLibrary/src/main/kotlin/com/vmware/aria/operations/Key.kt

```diff
@@ -0,0 +1,129 @@
+ * type must have the same set of identifiers.
+ */
+class Key @JvmOverloads constructor(
+    adapterType: String,
+    objectType: String,
+    name: String,
+    identifiers: List<Identifier> = emptyList()
+) {
+    val adapterType: String
+    val objectType: String
+    val name: String
+    private val identifiers: Map<String, Identifier>
+
+    init {
+        this.adapterType = adapterType
+        this.objectType = objectType
+        this.name = name
+        this.identifiers = HashMap()
+        for (identifier in identifiers) {
+            this.identifiers[identifier.key] = identifier
+        }
+    }
+
+    override fun toString(): String {
+        return "$adapterType:$objectType:$identifiers"
+    }
+
+    /**
+     * Return a list of data and identifiers that determine uniqueness
+     */
+    private val internalKey: List<Any>
+        get() {
+            // Sort all identifiers by 'key' that are part of uniqueness
+            val uniqueIdentifiers =
+                identifiers.values.filter { identifier -> identifier.isPartOfUniqueness }
+            val key = mutableListOf<Any>()
+            key.add(adapterType)
+            key.add(objectType)
+            if (uniqueIdentifiers.isEmpty()) {
+                //If there are no identifiers, or if all identifiers are not part of
+                // uniqueness, the name is used as uniquely identifying
+                key.add(name)
+            } else {
+                // Otherwise,if there is at least one identifier that is part of
+                // uniqueness, name is not used for
+                // identification. Add each of the unique identifiers to the list, sorted
+                // by key
+                key.addAll(uniqueIdentifiers.sortedBy { identifier -> identifier.key })
+            }
+            return key
+        }
+
+    /**
+     * @return True if the keys will resolve to the same object, false otherwise.
+     */
+    override fun equals(other: Any?): Boolean {
+        return if (other is Key) {
+            internalKey == other.internalKey
+        } else {
+            false
+        }
+    }
+
+    /**
+     * Return a hashcode of this Key, respecting the rules of key uniqueness
+     */
+    override fun hashCode(): Int =
+        internalKey.fold(0) { acc, id -> acc xor id.hashCode() }
+
+    /**
+     * Return the value for the given identifier key.
+     * @param key The identifier key.
+     * @param defaultValue An optional default value.
+     * @return The value associated with the identifier.
+     * If the value associated with the identifier is empty and 'defaultValue' is
+     * provided, returns 'defaultValue'.
+     * If the identifier does not exist, returns defaultValue if provided, else 'null'.
+    */
+    @JvmOverloads
+    fun getIdentifier(key: String, defaultValue: String? = null): String? {
+        val value = identifiers[key]?.value
+        return if (value?.isBlank() != false) {
```

Comment:
could simplified to 

```
return if (value?.isBlank()) {
```

Reply:
Actually, this is not exactly equivalent, because `value` is nullable. If it is, then `value?.isBlank()` evaluates to `null`, which Kotlin (and Java) will not coerce to a boolean.